src/useq/_grid.py

```diff
@@ -350,4 +362,126 @@ def _offset_y(self, dy: float) -> float:
         )
 
 
-AnyGridPlan = Union[GridFromEdges, GridRowsColumns, GridWidthHeight]
+# ------------------------ RANDOM ------------------------
+
+
+class Shape(Enum):
+    ELLIPSE = "ellipse"
+    RECTANGLE = "rectangle"
+
+
+class RandomPoints(_PointsPlan):
+    """Yield random points in a specified geometric shape.
+
+    Attributes
+    ----------
+    num_points : int
+        Number of points to generate.
+    max_width : float
+        Maximum width of the bounding box.
+    max_height : float
+        Maximum height of the bounding box.
+    shape : Shape
+        Shape of the bounding box. Current options are "ellipse" and "rectangle".
+    random_seed : Optional[int]
+        Random numpy seed that should be used to generate the points. If None, a random
+        seed will be used.
+    allow_overlap : bool
+        By defaut, True. If False and `fov_width` and `fov_height` are specified, points
+        will not overlap and will be at least `fov_width` and `fov_height apart.
+    """
+
+    num_points: int
+    max_width: float
+    max_height: float
+    shape: Shape
+    random_seed: Optional[int] = None
+    allow_overlap: bool = True
+
+    @property
+    def is_relative(self) -> bool:
+        return True
+
+    def __iter__(self) -> Iterator[GridPosition]:  # type: ignore
+        seed = np.random.RandomState(self.random_seed)
+        func = _POINTS_GENERATORS[self.shape]
+        n_points = max(self.num_points, MIN_RANDOM_POINTS)
+        points: list[Tuple[float, float]] = []
+        for x, y in func(seed, n_points, self.max_width, self.max_height):
+            if (
+                self.allow_overlap
+                or (None in (self.fov_width, self.fov_height))
+                or _is_a_valid_point(points, x, y, self.fov_width, self.fov_height)
+            ):
+                yield GridPosition(x, y, 0, 0, True)
+                points.append((x, y))
+            if len(points) >= self.num_points:
+                break
+        if len(points) < self.num_points:
+            _raise_warning(n_points, points)
+
+    def num_positions(self) -> int:
+        return self.num_points
+
+
+def _is_a_valid_point(
+    points: list[Tuple[float, float]],
+    x: float,
+    y: float,
+    min_dist_x: float | None,
+    min_dist_y: float | None,
+) -> bool:
+    """Return True if the the point is at least min_dist away from all the others.
+
+    note: using Manhattan distance.
+    """
+    if min_dist_x is None or min_dist_y is None:
+        return True
+    return not any(
+        abs(x - point_x) < min_dist_x and abs(y - point_y) < min_dist_y
+        for point_x, point_y in points
+    )
+
+
+def _raise_warning(n_points: int, points: list[Tuple[float, float]]) -> None:
+    """Raise a warning if the number of points is less than the requested number."""
+    warnings.warn(
+        f"Max number of iterations reached ({n_points}). "
+        f"Only {len(points)} points were found.",
+        stacklevel=2,
+    )
+
+
```

Comment:
don't make a new function `_raise_warning` just to raise a warning (unless that warning is going to be reused in multiple places)

Also, you don't need to check `if min_dist_x is None or min_dist_y is None` in `_is_a_valid_point` because it already happened in `__iter__`

 
```suggestion
        for x, y in func(seed, n_points, self.max_width, self.max_height):
            if (
                self.allow_overlap
                or self.fov_width is None
                or self.fov_height is None
                or _is_a_valid_point(points, x, y, self.fov_width, self.fov_height)
            ):
                yield GridPosition(x, y, 0, 0, True)
                points.append((x, y))
            if len(points) >= self.num_points:
                break
        if len(points) < self.num_points:
            warnings.warn(
                f"Unable to generate {self.num_points} points non-overlapping points."
                f"Only {len(points)} points were found.",
                stacklevel=2,
            )

    def num_positions(self) -> int:
        return self.num_points


def _is_a_valid_point(
    points: list[Tuple[float, float]],
    x: float,
    y: float,
    min_dist_x: float,
    min_dist_y: float,
) -> bool:
    """Return True if the the point is at least min_dist away from all the others.

    note: using Manhattan distance.
    """
    return not any(
        abs(x - point_x) < min_dist_x and abs(y - point_y) < min_dist_y
        for point_x, point_y in points
    )
```